Refactor and cleanup the influxdb_tools in traffic_stats

sbogacz opened the pull request:
1) Cleanup lint warnings in the traffic_stats/influxdb_tools go files (mostly due to repeated constant strings), and fix some idiomatic issues, like using both the  `+` operator and `fmt.Sprintf()` to concatenate strings on the same line.

2) Use the [errors package](https://github.com/pkg/errors) to report and add on to errors, rather than maintaining a global `errorMessage`.

3) Move the sync and create tools into separate folders so that the code can be tested (otherwise the compiler complains that main() and queryDB() are duplicated)

4) Use the [cli package](https://github.com/urfave/cli) to make the somewhat more user friendly. 

5) Test the `get*Stats` methods to ensure they behave accordingly, fix panic issues where a nil or empty slice gets passed in to those functions and the function doesn't check, and actually iterate over entire input slice, rather than only first result element

6) Add test file for traffic_stats.go, mostly to provide a starting point.

There's a few things going on in this PR, I got carried away and will have to open another subsequent PR for adding tests (and some refactoring if it's needed to make the code testable).

PSUdaemon commented on traffic_stats/influxdb_tools/create/create_ts_databases.go:
```diff
@@ -20,45 +20,81 @@ under the License.
+func queryDB(client influx.Client, cmd string, dbs ...string) (res []influx.Result, err error) {
+	db := ""
+	if len(dbs) > 0 {
+		db = dbs[0]
```
Why take a variadic argument if you will only ever use the first one? This seems like you want to have function overloading in a language that doesn't support it. It seems more idiomatic to pass a slice that can be empty.

sbogacz replied:
This is related to the changes I was hoping to add on top of this, in terms of trying to generalize the code currently in use to allow for code reuse (previously there were two separate functions, one allowing for `db` and one not, with the latter's body being identical except for instantiating the `influx.Query` passed an empty string instead. 

The change to a slice may well be better (but best would be the generalization of multiple queries then, and the reuse of common code).

PSUdaemon replied:
I'm all for reuse of code. There aren't any API contracts we need to adhere to. I'd prefer to just fix these outright.

sbogacz replied:
You'd prefer the fix in terms of sharing the code, or just the change from variadic to slice input? I'll be happy to make either (but moreso the former)

PSUdaemon replied:
Both. I think `traffic_stats.go` also uses a similar function.

PSUdaemon commented on traffic_stats/influxdb_tools/create/create_ts_databases.go:
```diff
@@ -20,45 +20,81 @@ under the License.
+)
+
+func createFlags() []cli.Flag {
+	return []cli.Flag{
```
Why clean up other constants, but put this into a function just to return a constant?

sbogacz replied:
That's definitely a premature optimization which I'd be happy to swap for a named var. Since there were no tests to start with, I had to A/B test most of the changes I was making. My "long" term goal was to abstract out something like a 
```go
type DBConfig struct {
    url string
    replication int
    user string
    password string
} 

func flagName(prefix, name string) string {
   if prefix == "" {
      return name
   }
   return fmt.Sprintf("%s-$s", prefix, name)
}

func envName(prefix, name string) string {
   if prefix == "" {
      return strings.ToUpper(name)
   }
   return strings.ToUpper(fmt.Sprintf("%s_$s", prefix, name))
}

func (c *DBConfig) Flags(prefix string) []cli.Flag {
	return []cli.Flag{
		cli.StringFlag{
			Name:  flagName(prefix, "url"),
                        EnvVar: envName(prefix, "url"),
			Usage: "The influxdb url and port",
			Value: "http://localhost:8086",
                        Destination: &c.url,
		},
		...
	}
}
```

Which would allow the start of some abstraction between the two commands, however, I realized part way that my changes were already larger than recommended, and that the course I was taking at the time (one executable with two commands) would actually break the build script as it stands.

PSUdaemon replied:
We aren't shy about taking contributions, so breaking it up into smaller more manageable chunks is preferred. That way we can more easily get code committed rather than trying to manage different facets of a large PR.

PSUdaemon commented on traffic_stats/influxdb_tools/create/create_ts_databases.go:
```diff
@@ -20,45 +20,81 @@ under the License.
 
 	influx "github.com/influxdata/influxdb/client/v2"
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
```
Why new deps?

sbogacz replied:
So as far as the [errors](https://github.com/pkg/errors) package is concerned, it's mostly to contextualize errors without doing something like 
```go
if err != nil {
   errorMessage = fmt.Sprintf("Some string %v", record) 
   fmt.Println(errorMessage)
}
```

This pattern was recurring in the code, and not only was it needlessly verbose (and using the global var errorMessage), but it also loses the original error, so debugging would entail going back to the code and adding Printf debugging, before the error itself could actually be addressed. The errors package functionality is actually quite simple, but like most simple code, it is very effective in what it does. 

The [cli](https://github.com/urfave/cli) package makes clis generally more user friendly, but also has some additional functionality (which I didn't add to this PR) of being able to set values on pointer receivers, declaring shell environment variables (helpful with containerized/orchestrated environment deployments).

As I was going through the codebase I was trying to come up with the different ways that it could be made more "idiomatic" (which I admit is a very vague term). Currently, some of the code seems more C/script-ish, and doesn't seems to take much advantage of the language beyond the syntax and garbage collecting.  This could well be because of constraints specified elsewhere, but if so, I didn't see anything that made me aware of them.

PSUdaemon replied:
`pkg/errors` actually seems pretty interesting. I saw it presented at the last GopherCon I think. The `cli` package seems to have some uses, but it looks like you aren't using them in the final form of this patch. I'd leave it out until it is necessary.

Both of these programs probably seem scripty because they were intended for limited use. One is for a one time setup of influx and I think the other was made for upgrades. They may have even once been written in Perl.

sbogacz replied:
The style makes sense, I was just trying to take the first pass at "Go-ifying" them. `pkg/errors` did get presented by Dave Cheney at the last GopherCon and I have to say that I have used it extensively, and it almost feels like an extension of the standard library.  I do agree that we can forego `cli` until absolutely necessary.

PSUdaemon commented on traffic_stats/influxdb_tools/sync/sync_test.go:
```diff
@@ -0,0 +1,230 @@
+
+	influx "github.com/influxdata/influxdb/client/v2"
+	"github.com/influxdata/influxdb/models"
+	. "github.com/smartystreets/goconvey/convey"
```
More new deps?

sbogacz replied:
The new dependency here is [goconvey](https://github.com/smartystreets/goconvey), which in my experience makes unit tests more accessible. Had there been an existing testing framework in use, I would have used that, but in lieu of one, I just used one that I consider to facilitate the creation of unit tests, since ease of use can sometimes drive adoption (so that the package gets better coverage).

PSUdaemon replied:
We could definitely use more testing. What does this get us over the built in testing framework?

sbogacz replied:
I'd say clarity of tests (rather than saying something like 
```go
if val != expected {
   t.Fail()
}
```
you can say something like 
```go
So(val, ShouldEqual, expectedVal)
```
), with several keywords (`ShouldContain`, `ShouldBeNil`, `ShouldNotBeNil`, etc.), as well as the availability of the context for testing, so that anonymous functions (say in a test httpHandler at test time) can make assertions as well. 

That said, the built-in test framework can definitely be used. My thought with the addition was mostly that it makes tests easier to reason about and maybe more approachable? If the aim is to minimize external dependencies, which I understand, I'd rewrite these with the standard testing library.

PSUdaemon replied:
I think @dewrich and @rob05c should weigh in here too. If using a framework like this means we get more people to contribute tests, I think that's something we should take into consideration too.

dewrich replied:
I've used GoConvey to monitor test coverage, but I'm not sure about the value add of tying the code to the framework.  I've typically just wrapped the Go testing framework with convenience methods when necessary.  We haven't had a larger discussion about these types of frameworks to weigh the pros/cons.

rob05c replied:
I generally agree with @PSUdaemon and @dewrich . Looking at goconvey, it seems to violate Separation of Concerns, blending infrastructure and build tools with testing code. Seems like those ought to be separate as possible; it should be possible to replace infrastructure tools or testing libraries independently. 

Don't get me wrong, the goconvey webpage stuff looks really useful, I'd certainly support a similar infrastructure tool which was independent of the test framework/library (ideally something integrating seamlessly with the native `go test`).

Also, in general, Go has built-in testing, is what goconvey adds really worth the cost of an external dependency?

All that said, if everyone else feels it will help testing _that_ much, I'm not extremely opposed. I vote "-0".

sbogacz replied:
I really do think that in general, having something that can handle assertions cleanly in some way is generally a lot mare palatable to read and write than just the straight `testing` package. Another alternative to goconvey could be [testify](https://github.com/stretchr/testify). I haven't used it much but it has some of the same advantages from goconvey, namely that instead of having something like
```go
if err != nil {
   t.Fail()
}
```
it becomes 
``` go
assert.NotNil(t, err)
```

This is even more nicer when operating on structured objects, because you can do deep comparisons, something like

```go
expectedResp := &influx.Response{ ... }
actualResp := influxClient.Query(...)
assert.Equal(t, actualResp, expectedResp)
```
which obviously cuts down the number of lines proportional to the number of fields you'd otherwise have to check manually.

Of course, we could go with neither framework, but my two cents in general are that making tests easy to add typically encourages people to write their own.

sbogacz replied:
@rob05c @dewrich @PSUdaemon Any thoughts on the above? I can undo `goconvey`, move it to `testify`, or just go with the more verbose testing option, but it'd be nice to only do the work once.

PSUdaemon replied:
I'm deferring to @rob05c and @dewrich on the testing.

sbogacz replied:
Sounds good. I'm working on moving some of the common influxdb code out, and creating some "autpopulating" config objects (with flag.Parse(), anyways).

dneuman64 replied:
IMO we should just use the standard testing package for this code.  This code is really more a set of loosely related scripts than a "program" which is why there wasn't any unit tests to begin with. Additionally, the logic in this code isn't changing all that often so encouraging people to write their own tests is not a concern to me.  If we want to explore the idea of using a test framework for our unit tests I think the Traffic Ops client code or goTM (Traffic_monitor/experimental) are better candidates.

rob05c replied:
I agree with using standard testing. The assertions @sbogacz mentioned are relatively easy to write. If we wanted, we could write a ~100 line library, with e.g. `func AssertEQ(a interface{}, b interface{}, t *testing.T) { if a != b { t.Errorf("expected %v == %v", a, b) } }`. I wouldn't be opposed to such a package, and maintaining that 100-line library is less expensive long-term than making all our tests depend on a third-party library that could stop being maintained at any time.

I'm not universally opposed to dependencies; but again, I don't see the value outweighing the cost here.

sbogacz commented:
Hey @PSUdaemon, thanks for the review. I'm more than happy to make changes to the PR to conform to project standards. I think the question of dependencies is a good one, both in terms of introducing new dependencies, but also in terms of versioning. As it stands, someone checking out this code would be using whatever version of the dependencies that they have checked out (or the version they checked out as they were trying to get the project to build). Vendoring is not the most fun topic, but it does solve a very real problem.

PSUdaemon commented:
I think we should at least start using the go vendoring functionality and see where that takes us. We already run into issues with our influxdb deps. We also have to be careful of licensing. I *think* all the licenses here are ok to bundle with an Apache project.

How about we start with the `pkg/errors` dep and then maybe @dewrich can provide some input of what he thinks of goconvey. I'm not anti-dependency, I just think we should be choosey.

By the way, are you using this code in production somewhere? We are always interested in hearing how this project is being used.

sbogacz commented:
I did look at the licenses beforehand, but I'll admit I don't know the most about the technicalities of Open Source licenses. I'll go back and get the tests to use the stdlib, as well as the cli flags. I would also advocate the use of vendoring... that said, if I remember correctly, there was statement from the godep author saying he was no longer maintaining the code, since there is now a team actually working on a "native" vendoring solution.

I'm not currently using this code in production, I was pointed to it recently and curious about it, so as I was taking a look I figured I'd try to help out.

dneuman64 commented:
+1 on vendoring.  We are currently defining the version of the InfluxDb client to use in the build script and vendoring is a much better solution.

sbogacz commented:
I've had good experiences with `godep` for the most part, and I'd be happy to bring it in (I can maybe throw in a Makefile to make some of the local commands easier). That said, [a native solution is in the works](https://groups.google.com/forum/#!msg/go-package-management/P8TehVoFLjg/Ni6VRyOjEAAJ) so it will probably have to be updated at some point.

rob05c commented:
+1 on built-in vendoring, -1 on godep unless there's a compelling reason.

I'd also be +1 on Makefiles iff there's a compelling reason for them. Go build tools are minimal, but I've only found the need for Makefiles in about 10% of projects. If we only need a couple build lines, e.g. a `go generate;go build` or a complex command like `go build -ldflags "-X main.GitRevision=`git rev-parse HEAD` -X main.BuildTimestamp=`date +'%Y-%M-%dT%H:%M:%S'`"`, Make isn't worth the complexity over simple two-line `build.sh`. IMO. But definitely Make over something less standard.

sbogacz commented:
@rob05c Sadly, there's not actually a native vendoring tool solution yet, and I haven't heard any news about when it could be expected. While we can certainly move dependencies into a vendor/ directory manually, that can be a slippery slope, since you need to also vendor the dependencies of your dependencies, which can conflict with the versioning of your own dependencies. There's a few rather annoying scenarios for this, and having some form of tooling around it generally helps. On the other hand, there are currently relatively few dependencies in the project... that said, as the Go codebase grows, the more intractable the manual solution becomes.

rob05c commented:
@sbogacz Have you seen https://golang.org/cmd/go/#hdr-Vendor_Directories ?

sbogacz commented:
@rob05c Yeah, that's what I'm talking about. That functionality (and since greater versions of go1.6, the tooling as well) supports the _existence_ of a vendor folder, and knows to look in there before defaulting to the GOPATH to find an import in the code. In terms of actually _building_ the vendor folder, is where there's no native solution (probably tooling) yet. For example, if we take two hypothetical libraries, both of which maybe provide a layer over NSQ to do cache invalidation for a distributed cache. One version may be using the new standard library `context` while the other is still using `golang.org/x/net/context`. 

Since many libraries don't actually vendor (probably rightly so most of the time), it would then be up to whoever is saving that dependency to get the separate contexts in a sub-vendor folder within the vendor tree. It's certainly not impossible, but it's a problem that scales with the number of dependencies. What tools like `godep` and `govendor` do is to automatically navigate the GOPATH looking for the dependencies (even figuring out peculiarities like symlinked directories etc.), and try to build a vendor tree with the flattest structure possible, so that you don't have say, 18 copies of "google.golang.org/grpc/" (for example). Does that make sense, I'm not sure that's the best explanation.

Maybe the better way to put it is this: Go supports vendoring, in the sense of the existence of the vendor directory, and has since Go1.6 (and experimentally in 1.5). `godep` or `govendor` are tooling options to allow for the automated building of vendor directories.

rob05c commented:
Right.

> On the other hand, there are currently relatively few dependencies in the project... that said, as the Go codebase grows, the more intractable the manual solution becomes.

Right now, Traffic Control doesn't have enough Go dependencies for sub-dependency conflicts to be an issue, and in my experience Go idioms don't tend toward that many dependencies. So, I think YAGNI applies, and I vote we keep using the built-in vendoring until sub-dependency conflicts become an issue, and reconsider a third-party tooling dependency then.

sbogacz commented:
Added some more commits to show the direction I was thinking of going with some of the refactoring, what the move back to `flag` from `urfave/cli` looks like, and added the `vendor` directory. These changes so far work fine locally. If these look OK to you guys @PSUdaemon @rob05c @dewrich, then I'll propagate them to the sync command, and traffic_stats.go.

I'll hold off on changing the tests until there is consensus.

sbogacz commented:
@dneuman64 These last changes should encapsulate what we talked about:

* Remove `urfave/cli` dependency by having a config object that can be "autopopulated" by the standard library flag
* Make create and sync commands use the config object
* Add some test functions (under package name assert, both for reuse between the tool tests, and the eventual tests for traffic_stats), and rewrite the sync tests to use them, removing the `goconvey` dependency

The create and sync scripts still run locally (for me), and nothing has really changed for traffic_stats.go.

dneuman64 commented:
thanks @sbogacz, I will test it all out when I get a chance and let you know if I find any issues.

rob05c commented on traffic_stats/assert/testhelpers.go:
```diff
@@ -0,0 +1,109 @@
+			t.Errorf("expected %v to be empty, but it is not", a)
+		}
+	case reflect.String:
+		if val.Len() != 0 {
```
Since these are all identical, all but the last one can be replaced with `fallthrough`

rob05c replied:
FYI not rejecting, just noting for future improvements.

sbogacz replied:
That would be good cleanup. I'm going to refrain from doing it in this PR, since I think it's probably grown enough as is.

rob05c commented on traffic_stats/assert/testhelpers.go:
```diff
@@ -0,0 +1,109 @@
+// Equal takes two values, checks whether they are equal (using a deep equal
+// check) and fails the test if not
+func Equal(t *testing.T, a, b interface{}) {
+	if !reflect.DeepEqual(a, b) {
```
DeepEqual can be slow, and might be unexpected, callers might expect 'Equal' to actually check '=='. Might be worth making a separate `func Equal` and `func DeepEqual`

sbogacz replied:
True. Do you think it's worth making the change now, or as a follow up PR?

rob05c replied:
Follow up is fine, just making the observation.

dneuman64 commented:
It looks like there are some errors in the vendored InfluxDb code when I try to build traffic_stats. I will dig around a little bit and see what I can figure out, unless this rings a bell with you?
```
github.com/apache/incubator-trafficcontrol/traffic_stats/vendor/github.com/influxdata/influxdb/client/v2
# github.com/apache/incubator-trafficcontrol/traffic_stats/vendor/github.com/influxdata/influxdb/client/v2
vendor/github.com/influxdata/influxdb/client/v2/client.go:302: undefined: models.NewTags
vendor/github.com/influxdata/influxdb/client/v2/client.go:329: p.pt.Tags().Map undefined (type models.Tags has no field or method Map)
vendor/github.com/influxdata/influxdb/client/v2/client.go:344: not enough arguments to return
vendor/github.com/influxdata/influxdb/client/v2/udp.go:78: p.pt.Round undefined (type models.Point has no field or method Round)
vendor/github.com/influxdata/influxdb/client/v2/udp.go:79: p.pt.StringSize undefined (type models.Point has no field or method StringSize)
vendor/github.com/influxdata/influxdb/client/v2/udp.go:85: p.pt.AppendString undefined (type models.Point has no field or method AppendString)
vendor/github.com/influxdata/influxdb/client/v2/udp.go:90: p.pt.Split undefined (type models.Point has no field or method Split)
++ pwd
+ echo 'Could not build go program at /home/dneuma4602k/incubator-trafficcontrol/rpmbuild/BUILD/traffic_stats-1.9.0: '
Could not build go program at /home/dneuma4602k/incubator-trafficcontrol/rpmbuild/BUILD/traffic_stats-1.9.0:
+ exit 1
error: Bad exit status from /var/tmp/rpm-tmp.KyDW2S (%build)
```

dneuman64 commented:
The influxdb client has a depedency on `github.com/influxdata/influxdb/models` so that directory needs to be added to the vendor folder.

sbogacz commented:
That's the difficulty in vendoring without tooling, totally missed those. I'll add to the vendor folder and push

sbogacz commented:
@dneuman64 I had to bring in several more dependencies into vendor to cover those dependencies' dependencies (not to beat a dead horse, but tooling can really help here). I won't have a good chance to test anytime soon, but let me know if that fixes the issue, or if there's still a problem.

dneuman64 commented on traffic_stats/build/traffic_stats.spec:
```diff
@@ -75,8 +75,8 @@ godir=src/github.com/apache/incubator-trafficcontrol/traffic_stats/influxdb_tool
-  go build sync_ts_databases.go
-  go build create_ts_databases.go
+  go build sync/sync_ts_databases.go
+  go build create/create_ts_databases.go
```
It won't let me comment on line 69, but it needs to be removed since we are vendoring now.
`go_get_version "$oldpwd"/src/github.com/influxdata/influxdb v1.0.1 && \`

sbogacz replied:
@dneuman64 Removed and pushed

dneuman64 commented:
ok, I think just one more thing.  The vendored traffic_ops/client code is out of date.  Can you get the latest master version of client code in there?  Might be best to just rebase onto the latest master as well. As is stands now traffic stats won't work with a traffic ops running master.

sbogacz commented:
@dneuman64 Crossing fingers

dneuman64 commented:
Looks like you still need to get the latest traffic_ops/client files into traffic_stats/vendor

sbogacz commented:
🤦‍♂️  Pulled in upstream but didn't update vendor, so here we go again...

dneuman64 commented:
👍 Everything looks good now. 
One thing that does need to be done is to update the docs to reflect the new flag names for the influxdb tools.  I am still going to merge this PR, but it is something that needs to be done, either by us or by you.

http://trafficcontrol.apache.org/docs/latest/admin/traffic_stats.html#id2

sbogacz commented:
@PSUdaemon @dneuman64 @rob05c @dewrich Hey guys, in case you haven't seen it yet, [they just released the pre-alpha version of the native dependency tool for Go](https://groups.google.com/forum/#!topic/go-package-management/et1qFUjrkP4/discussion). [Here it is](https://github.com/golang/dep).

I don't think it's ready for use just yet, but I wanted to put it on your radar.
